### PR TITLE
Handle LTI launch error due to browser cookie settings (#312)

### LINKED ISSUE #312: Error when launching CCM in Safari
When launching CCM from Safari on Canvas beta this error returns

`{"status":401,"error":"Unauthorized","details":{"description":"Error validating ltik or IdToken","message":"MISSING_VALIDATION_COOKIE"}}`

I believe this is a common issue with Safari if a user has the default Privacy option "Prevent cross-site tracking" checked. I saw it was [brought up as an issue in the library](https://github.com/Cvmcosta/ltijs/issues/53) but there was no resolution. The resultion by most other vendors is either turn this setting off or switch to a vanity URL, which we've been in the process of doing but isn't complete.

I tested on Canvas Test which does have a vanity URL and it works fine. However this will be the same in prod as it is in beta.

Our python apps that use pylti13 prompt to open a new window. That's because there's [javascript to check for this](https://github.com/dmitry-viskov/pylti1.3/blob/master/pylti1p3/cookies_allowed_check.py) in the library. Since the majority of our users are on Safari this might be something we need to address and at least prompt users to open in a new window or disable this option. 

IMS LTI team is working on a solution to avoid this by using postMessage rather than cookies but this isn't yet in the spec.

### PR DESCRIPTION
**Pull request by ssciolla**
The PR aims to resolve #312.

Screenshot of error page

Screen Shot 2022-03-09 at 3 59 39 PM

### CODE REVIEW

File: ccm_web/server/src/lti/lti.service.ts [hunk 1]
```diff
@@ -113,6 +113,10 @@ export class LTIService implements BeforeApplicationShutdown {
       })
     })
 
+    provider.onInvalidToken(async (req: Request, res: Response) => {
+      return res.redirect('/launch-error')
```
Comment (pushyamig) [inline ccm_web/server/src/lti/lti.service.ts:117]: Can you tell the difference between launch from safari vs student launching CCM (by somehow grabbing url)? doesn't it both invalid token launches. 

What you have right now will work and not asking to detect safari, just want to know if there is a way LTIJS is giving that info.

Comment (ssciolla) [reply ccm_web/server/src/lti/lti.service.ts:117]: I think if it's a student, the launch still works, we just don't log them in and redirect them to `access-denied`. See `onConnect` hook.

File: ccm_web/client/src/pages/LaunchError.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+export default function LaunchError (): JSX.Element {
+  const classes = useStyles()
+
+  return (
+    <div className={classes.container}>
+      <InlineErrorAlert>
+        <Typography variant='subtitle1' component='h1'>Launch Error</Typography>
+        <Typography>
+          The tool could not be launched. This is likely due to settings related to cookies on your Internet browser.
```
Comment (pushyamig) [inline ccm_web/client/src/pages/LaunchError.tsx:21]: Should the text be updated as per this https://github.com/tl-its-umich-edu/canvas-course-manager-next/issues/312#issuecomment-1055860902

Comment (ssciolla) [reply ccm_web/client/src/pages/LaunchError.tsx:21]: Okay, I'll do this today.

File: ccm_web/server/src/lti/lti.service.ts (hunk 1 above)
Comment (pushyamig) [reply ccm_web/server/src/lti/lti.service.ts:117]: I looked it up a bit more. I see that `LTI login/authorize/Launch` is happening if 3rd-party cookies are disabled. some how the code block `onInvalidToken` is detecting as cookies are disabled and tool launch is failed. So I did not see from the code either (documentation of LTIJS is not much about this) much description about this function. 

I would just go with an understanding that **onInvalidToken** is written for seeing cookies disable or enabled. IF that is not true then we can find out based on the error case that comes to us in future.

File: ccm_web/client/src/pages/LaunchError.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        <Typography>
+          The tool could not be launched. It is likely your browser is blocking third-party cookies,
+          which are required for this application.
+        </Typography>
+        <Typography>
+          To resolve this issue, please update your settings
+          for <Link href={chromeLink}>Chrome</Link>, <Link href={macOSLink}>Safari for MacOS</Link>
+          , <Link href={iOSLink}>Safari for iOS</Link>, <Link href={firefoxLink}>Firefox</Link>
+          , <Link href={edgeLink}>Microsoft Edge</Link>,
+          or another preferred browser to allow third-party cookies.
```
Comment (pushyamig) [inline ccm_web/client/src/pages/LaunchError.tsx:31]: So far no objection from Melinda about this text and on that note this is ok to merge

**Review by pushyamig — APPROVED** [review]

**Comment by ssciolla** [comment]
Thanks, @pushyamig, I'll merge and we can always adjust text later.

**Pull request merged**